genrl/agents/modelbased/cem/cem.py

```diff
@@ -0,0 +1,169 @@
+            for a in actions_batch[i]
+        ]
+
+        return torch.cat(elite_states, dim=0), torch.cat(elite_actions, dim=0)
+
+    def select_action(self, state):
+        state = torch.as_tensor(state).float()
+        action, dist = self.agent.get_action(state)
+        return action, torch.zeros((1, self.env.n_envs)), dist.log_prob(action).cpu()
+
+    def update_params(self):
+        sess = [self.plan() for _ in range(100)]
+        batch_states, batch_actions, batch_rewards = zip(*sess)
+        elite_states, elite_actions = self.select_elites(
+            batch_states, batch_actions, batch_rewards
+        )
+        action_probs = self.agent.forward(elite_states.float())
+        loss = F.cross_entropy(
+            action_probs.view(-1, self.action_dim),
+            elite_actions.long().view(-1),
+        )
+        self.logs["crossentropy_loss"].append(loss.item())
+        loss.backward()
+        # torch.nn.utils.clip_grad_norm_(self.agent.parameters(), 0.5)
+        self.optim.step()
+
+    def get_traj_loss(self, values, dones):
+        # No need for this here
+        pass
+
+    def collect_rollouts(self, state: torch.Tensor):
+        """Function to collect rollouts
+
+        Collects rollouts by playing the env like a human agent and inputs information into
+        the rollout buffer.
+
+        Args:
+            state (:obj:`torch.Tensor`): The starting state of the environment
+
+        Returns:
+            values (:obj:`torch.Tensor`): Values of states encountered during the rollout
+            dones (:obj:`torch.Tensor`): Game over statuses of each environment
+        """
+        states = []
+        actions = []
+        for i in range(self.rollout_size):
+            action, value, log_probs = self.select_action(state)
+            states.append(state)
+            actions.append(action)
+
+            next_state, reward, dones, _ = self.env.step(action)
+
+            if self.render:
+                self.env.render()
+
+            self.rollout.add(
+                state,
+                action.reshape(self.env.n_envs, 1),
+                reward,
+                dones,
+                value,
+                log_probs.detach(),
+            )
+
+            state = next_state
+
+            self.collect_rewards(dones, i)
+
+            if dones:
+                break
+
+        return states, actions
+
+    def collect_rewards(self, dones: torch.Tensor, timestep: int):
+        """Helper function to collect rewards
+
+        Runs through all the envs and collects rewards accumulated during rollouts
+
+        Args:
+            dones (:obj:`torch.Tensor`): Game over statuses of each environment
+            timestep (int): Timestep during rollout
+        """
+        for i, done in enumerate(dones):
+            if done or timestep == self.rollout_size - 1:
+                self.rewards.append(self.env.episode_reward[i].detach().clone())
+                # self.env.reset_single_env(i)
```

Comment:
Why is this commented out? This is necessary to reset environments immediately as they are set to done. (Not a good practice to do `env.step()` if the env is already returning `done = True`)

Reply:
Since I am breaking the loop of actions if a `env.step()` returns `done=True`, and every plan session (the `plan` function) starts with `env.reset()`, I think this is redundant here, hence its commented out